### PR TITLE
ToolsPanel: switch to plus icon when no controls present in panel body

### PR DESCRIPTION
**Pull request by ramonjd**
**Depends on:** #34530 

Today we're swapping the ellipsis icon with a plus symbol when there are no selected controls within a ToolsPanel. 

Screen Shot 2021-08-17 at 8 03 46 pm

The [original suggestion](https://github.com/WordPress/gutenberg/pull/34060#issuecomment-899315955). Thank you!

The motivation is to provide a hint to users that they can add optional controls to the panel. 

"Optional" controls – introduced in https://github.com/WordPress/gutenberg/pull/34530 – are those not displayed by default, and whose visibility may be toggled.

Some known edge cases/considerations:

- When the ToolsPanel has children that are not rendered (and therefore selectable) in the panel menu. Not sure yet, but we might have to know whether there are children that not registered ToolsPanelItems.

Several ToolsPanels will therefore look something like this:

Screen Shot 2021-08-17 at 8 18 38 pm

## Testing

1. Fire up this branch, and insert a group, cover or other block that has a dimensions ToolsPanel in the sidebar controls.
2. If there is no optional control present in the panel body, the plus icon should display in the ToolsPanel header. "Default" controls cannot be toggled and should have no effect on the icon.
3. Toggle controls on and off in the ToolsPanel menu. The ellipsis should show when there are controls present in the panel body.
4. Check that panel with only default controls, and no optional controls, always show the vertical dots icon.
5. Check the `aria-label` for both button states. They should reflect the label prop passed down to the dropdown component.

Run ` npm run test-unit packages/components/src/tools-panel/test/index.js` for glory!

## Screenshots 

https://user-images.githubusercontent.com/6458278/129707348-74654137-21f5-4bd4-bd8c-0e310333fda6.mp4

New view in storybook:

Screen Shot 2021-09-15 at 10 40 03 pm

## Types of changes

UI: Adding an icon to the ToolsPanel header.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/tools-panel/tools-panel-header/hook.js
Comment (andrewserong) [inline packages/components/src/tools-panel/tools-panel-header/hook.js]: Looks like prettier is complaining that this is all on one line. Locally, this seems to do the trick:

```suggestion
	const hasSelectedMenuItems = menuItemsArray.some(
		( [ , isSelected ] ) => isSelected
	);
```

File: packages/components/src/tools-panel/tools-panel-header/component.js
Comment (andrewserong) [inline packages/components/src/tools-panel/tools-panel-header/component.js]: Tiniest of tiny nits: the other identifiers in this list (aside from the one below) are in alphabetical order. Shall we move this one up to just beneath `hasMenuItems`?

**Review by andrewserong — COMMENTED** [review]
It's a small change this one, but such a nice improvement @ramonjd! Looks great:

tools-panel-plus-icon-sml

Looks like the linting CI step is failing because of prettier complaining about line length, but other than that, LGTM!

**Review by andrewserong — APPROVED** [review]
The linter looks happy now, and it's still working well. Thanks for updating, LGTM! 🚀

**Comment by ramonjd** [comment]
> Looks like the linting CI step is failing because of prettier complaining about line length, but other than that, LGTM!

Thanks for testing @andrewserong My local linter has stopped pulling me up on this for some reason pre-commit 💀 

> When the ToolsPanel has children that are not rendered (and therefore selectable) in the panel menu. Not sure yet, but we might have to know whether there are children that not registered ToolsPanelItems.

My only idea so far on how to get around this was to check the immediate children of `<ToolsPanel />` for components that are not of type `ToolsPanelItem` (and therefore not registered to appear in the header toolbar).

So, in the tools-panel/hook, something like:

```js

	// Checks for the existence of children that are not ToolPanelItems.
	const [ hasOtherChildren, setHasOtherChildren ] = useState( false );

	useEffect( () => {
		const otherChildren = Children.toArray( props.children ).filter(
			( { type } ) => type?.displayName !== 'ToolsPanelItem'
		);

		setHasOtherChildren( otherChildren.length > 0 );
	}, [ props.children ] );
```

which we'd add to context and access in the tools header to make our decision.

Needs some more testing, so not sure whether it should be part of this PR or not.

**Comment by andrewserong** [comment]
> Needs some more testing, so not sure whether it should be part of this PR or not.

Good question! The fallback behaviour of displaying the three dots if there are non panel item children might not be too bad... unless it gets tripped up once the [SlotFill](https://github.com/WordPress/gutenberg/pull/34063) is added, in which case it might be that the plus icon never gets shown (if the Slot gets counted) 🤔

I think because the fallback behaviour is to show the three dots (like we do currently), I'd personally lean toward going with this and tweaking in follow-ups as we encounter issues. But happy to go either way — apologies for the early/eager review if you want to take a bit longer to explore your options! 😀

**Comment by ramonjd** [comment]
> But happy to go either way — apologies for the early/eager review if you want to take a bit longer to explore your options!

Not at all! Very glad for the feedback. 🙏

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for putting this together @ramonjd. 

These sorts of improvements definitely help polish the component ✨ 

All in all, this tested well for me. 

✅ Displays plus icon when its initial state is empty
✅ Dots icon displays when its initial state contains visible controls
✅ Icon correctly updates when controls are toggled on/off

The new plus icon looks good. At first, I wasn't sure how it might play when all the controls in the `ToolsPanel` have been toggled off and the user wants to "reset all". In this situation, the "reset all" option would only make a difference if it meant restoring the display of default controls. In that instance, those controls are "added" back into the panel, so the plus icon still makes some sense.

> My only idea so far on how to get around this was to check the immediate children of <ToolsPanel /> for components that are not of type ToolsPanelItem (and therefore not registered to appear in the header toolbar).

This approach while helpful could also be a bit brittle. The `ToolsPanelItem` sub-component and the `ToolsPanel`'s menu were built in such a way that a `ToolsPanelItem` could be wrapped or grouped in another component. This might make the check for "not of type `ToolsPanelItem` a little complicated.

If pursuing this we probably need to ensure we can eliminate false positives from cases such as:

```
const WrappedItem = ( props ) => {
	return (
		<div>
			<ToolsPanelItem { ...props }>
				<div>Just to be difficult.</div>
			</ToolsPanelItem>
		</div>
	);
};

const GroupedItems = () => {
	return (
		<>
			<ToolsPanelItem>
				<div>Grouped control</div>
			</ToolsPanelItem>
			<ToolsPanelItem>
				<div>Alt grouped control</div>
			</ToolsPanelItem>
		</>
	);
};
```

It's hard to see a case where a control would be added to the panel that isn't within the menu. We might add things such as help text or contrast checkers but I'm not sure that would remove the benefit of having the plus icon to encourage the user to explore the more advanced settings available through the menu.

I'll be interested to hear the thoughts of more seasoned UI gurus.

**Review by apeatling — COMMENTED** [review]
I've tested this and it works as expected. I think the plus icon only when no items have been selected or are present by default is a good approach.

I think @shaunandrews might want to chime in here since he designed something with the plus menu present all of the time. Personally I'm not sure about this because it means hitting a plus button when removing/unchecking an item -- which seems strange.

**Review by stacimc — COMMENTED** [review]
This looks good and tests well for me :)

> unless it gets tripped up once the SlotFill is added, in which case it might be that the plus icon never gets shown

Just echoing this -- it looks like this might be the case? If so, I think the plus icon would never get shown for the Dimensions panel once #34063 is merged 🤔 Depending on how likely that is to land soon, maybe that changes the priority of handling the edge case. Particularly if the SlotFill isn't added to all panels, it would be odd for them to behave differently.

Otherwise, falling back to the three dots seems fine to me.

**Comment by shaunandrews** [comment]
image

I tried a bunch of different icons. It might sound strange, but I almost think we could keep the chevron icon:

image

But I think it might be really confusing to use the same icon as the accordion, but open a menu instead. Though, to be fair, we use the chevron icon in _many_ places to indicate a menu.

--

I did suggest to @apeatling that we might consider using the same icon regardless of the presence of tools. I did try having the icon change once a tool was added, but I found myself looking at a strange pattern of icons (vertical dots, , cross, horizontal dots, plus, arrow) down the right side of my browser, which felt visually unpleasant.

image

--

When there are no tools present, I think the menu should omit the "Reset all" option:

image

**Comment by ramonjd** [comment]
Thanks for the feedback everyone!

> I think the plus icon would never get shown for the Dimensions panel once #34063 is merged 🤔 Depending on how likely that is to land soon, maybe that changes the priority of handling the edge case. Particularly if the SlotFill isn't added to all panels, it would be odd for them to behave differently.

Thanks for pointing this out. I haven't tested it on top of #34063 (if that's the direction we're taking with slots). I'll check it out and see if we can handle it.

My opinion is that this UI change is a nice to have, so I don't there's harm holding off until the ground is stable.

> The ToolsPanelItem sub-component and the ToolsPanel's menu were built in such a way that a ToolsPanelItem could be wrapped or grouped in another component. This might make the check for "not of type ToolsPanelItem a little complicated.
> ...
> It's hard to see a case where a control would be added to the panel that isn't within the menu. We might add things such as help text or contrast checkers but I'm not sure that would remove the benefit of having the plus icon to encourage the user to explore the more advanced settings available through the menu.

Thanks for confirming. I admit, it was a fairly shallow assessment. Looking at the ToolsPanels tests reveals more to me how it's intended to work.

I agree that we should be trying to help the user as much as we can to uncover hidden controls. The plus symbol to me is probably the least intrusive and least elaborate way to achieve this.

Given the small footprint of this PR, I'd be fine with merging it and then monitoring the [discussion](https://github.com/WordPress/gutenberg/pull/34063) surrounding slotfills.

I'll get around to testing this on top of that branch at least soon.

> I did try having the icon change once a tool was added, but I found myself looking at a strange pattern of icons (vertical dots, , cross, horizontal dots, plus, arrow) down the right side of my browser, which felt visually unpleasant.

I agree with this observation. I think icon changes might make more sense, or rather, testing icon changes makes more sense, after we migrate more of the block supports over to the ToolsPanel. 

It's probably just me , but the ⚙️ symbol speaks more to me than a vertical ellipsis, but my primate brain will try to click on anything so the outcome is the same. 

I suppose users won't spend too much time worrying about the content of the icon before their synapses conspire to compel the finger/hand to click on it.

**Comment by ramonjd** [comment]
Holding off until the changes in **ToolsPanel: Refine component behaviour** https://github.com/WordPress/gutenberg/pull/34530 are finalized

**Comment by aaronrobertshaw** [comment]
The PRs previously blocking this have been merged:
- https://github.com/WordPress/gutenberg/pull/34157
- https://github.com/WordPress/gutenberg/pull/34530

This will need a rebase. While it's being updated, would it be a good idea to add a new example to the component's storybook entry demoing the panel in an initially empty state?

**Comment by ramonjd** [comment]
> This will need a rebase. 

Done! Thank you! 

I've also tweaked the logic to only take into account optional menuItems, since I couldn't think of a better way to apply the original intent behind this PR to the new UX.

I had plans to create tests for the [ToolsPanel hook](https://github.com/WordPress/gutenberg/blob/ce321f22303fc19af150bf530338dab2b9768095/packages/components/src/tools-panel/tools-panel/hook.js), but ran out of time today.

> While it's being updated, would it be a good idea to add a new example to the component's storybook entry demoing the panel in an initially empty state?

Definitely, thanks for the idea.

I did notice that the default story already displayed the plus icon, since the min-height isn't shown by default (and therefore optional)

Screen Shot 2021-09-15 at 10 31 00 pm


So I made them all display by default, and created a story specifically to showcase the plus icon.

The default now looks like this:
Screen Shot 2021-09-15 at 10 30 28 pm

And the new story:

Screen Shot 2021-09-15 at 10 40 03 pm

What do you think? I'm happy to revert it since the default story demonstrated the plus icon anyway.

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel/hook.js]: Nit:

Since `Array.some(...)` works fine on an empty array fine, the `optionalMenuItemsArray.length > 0` check is not necessary.

This snippet could be simplified to

```suggestion
			setAreOptionalControlsHidden(
				! Object.entries( menuItems.optional ).some(
					( [ , isSelected ] ) => isSelected
				)
			);
```

**Review by ciampo — COMMENTED** [review]
> I had plans to create tests for the ToolsPanel hook, but ran out of time today.

This would be great. I was also thinking about having unit tests that interact with the `ToolsPanel` component, showing/hiding optional items and checking that the expected icon is being shown.

About the Storybook examples — what if we kept the "Minimum height" panel in the "Default" story as optional, but we made it visible (if that's even possible)? This should still hide the plus icon on the first initial render, but it will make for a more complete example (since this is the default one). Hopefully I'm making sense 😅

Also, another thing that I find useful when browsing Storybook examples is if the text in the examples describes what is being showcased. For example, the `label` props for each `ToolsPanel` and `ToolsPanelItem` component could read something like this:

Screenshot 2021-09-15 at 15 30 48

Screenshot 2021-09-15 at 15 32 00

Just a thought, though :)

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (ramonjd) [reply packages/components/src/tools-panel/tools-panel/hook.js]: Thanks for the suggestion. The result of the above code change would return `true` when there are no optional defaults.

We want to track whether optional items are available at all first, then, if so, check whether are they hidden.

That means we also want to return `false` for the case where there are no optional items. Where there are only "default" items, no controls are toggable^ (is that a word?) 😄 

Unfortunately I named the var and set method a little too opaquely to make this clear, so my bad.  It probably should be something like `setAreOptionalControlsAvailableAndHidden`, which is a mouthful, but not as long as some Welsh train station names.

^ _Toggly_ or _togglesworth_ seem like nice alternatives.

**Comment by ramonjd** [comment]
Thanks for reviewing @ciampo !

> Also, another thing that I find useful when browsing Storybook examples is if the text in the examples describes what is being showcased. For example, the label props for each ToolsPanel and ToolsPanelItem component could read something like this:

This was a late night copy 🍝 . Thanks for catching that one!

> I was also thinking about having unit tests that interact with the ToolsPanel component, showing/hiding optional items and checking that the expected icon is being shown.

True, I looked briefly at the DOM nodes generated and I couldn't find a neat react testing library hook (role, text etc) to check the icon toggling states unless I update the `aria-label` to reflect the state of the icon. This is because the icon attributes, aside from the SVG attribute values, aren't unique.

Therefore I thought I'd start with the hooks before making a mess of the component tests. I'm not an expert in React Testing Library so if there are any tricks folks would like to share, I'm ready to learn!

Now I think about it, might not be a bad idea to adjust the label, at least to make it more descriptive. Currently the `aria-label` for the ToolsPanel icon is `"Dimensions"`, which repeats the encasing `h2` text anyway.

Screen Shot 2021-09-16 at 8 40 52 am

So maybe something like `aria-label="View options"` for the vertical ellipsis, and `aria-label="View and add options"`

I'm very certain others can come up with better labels than that! I'm just proposing the toggling of labels. 😆

**Comment by ramonjd** [comment]
> what if we kept the "Minimum height" panel in the "Default" story as optional, but we made it visible (if that's even possible)? This should still hide the plus icon on the first initial render, but it will make for a more complete example (since this is the default one). Hopefully I'm making sense 

I get it! This is a good idea, thanks. I suppose it might require sending a flag to the ToolsHeader (maybe). I'm not sure yet, but I'll check it out.

_Update_: I'm not sure about this. The DropDownMenu has no prop to toggle initial visibility (as far as I can see), and I can't find anything useful in the StoryBook API either, though I only looked for a minute or two so it might not count for much :)

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (ramonjd) [reply packages/components/src/tools-panel/tools-panel/hook.js]: You got me thinking though, what if we show the plus icon _as long as there are hidden options left to "unhide"_, e.g.,

plus-icon-every

Then we could use `every`:

```js
			setAreOptionalControlsAvailableAndHidden(
				! Object.entries( menuItems.optional ).every(
					( [ , isSelected ] ) => isSelected
				)
			);
```

**Comment by ramonjd** [comment]
> Now I think about it, might not be a bad idea to adjust the label, at least to make it more descriptive. Currently the aria-label for the ToolsPanel icon is "Dimensions", which repeats the encasing h2 text anyway.

I've done this in 0211dbe51100bee704241488f0c8a12db2a1c02b as a tentative commit :D

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (ciampo) [reply packages/components/src/tools-panel/tools-panel/hook.js]: Thank you @ramonjd for detailing the logic — my bad, I had missed some parts of it!

> what if we show the plus icon as long as there are hidden options left to "unhide",

That is also a good alternative. Both approaches seem reasonable — _maybe_ the original approach seems to make more sense to me (but it may be just because I already got accustomed that logic?)

> but not as long as some Welsh train station names

Yeah, not sure if we should call our variables `Llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch`

File: packages/components/src/tools-panel/tools-panel-header/component.js
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-header/component.js]: Since the menu isn't only used to "show" and "add" option, but also to "hide", "remove" and "reset" them, maybe we should use a term that is different from all the ones I've just listed. I think your initial suggestion of "View" may be a bit safer — and I also thought about "Manage" as an alternative.

Another quick thought — what if, instead of "options", we used  "sections"?

To sum it up, a few alternatives could be:

- "Manage sections" / "Manage and add sections"
- "Manage options" / "Manage and add options"
- "View sections" / "View and add sections"
- "View options" / "View and add options"

File: packages/components/src/tools-panel/test/index.js
Comment (ciampo) [inline packages/components/src/tools-panel/test/index.js]: You can use `.not.toBeInTheDocument()` instead of `.toBeNull()`

**Review by ciampo — COMMENTED** [review]
> I'm not sure about this. 

Yeah, you're actually correct — I don't think that currently a `ToolsPanelItem` can be set as optional but initially visible.

> Now I think about it, might not be a bad idea to adjust the label, at least to make it more descriptive.

That is a good idea!

> I'm very certain others can come up with better labels than that! I'm just proposing the toggling of labels. 😆

Naming and labeling is hard! I've left my suggestions in an inline comment

> Therefore I thought I'd start with the hooks before making a mess of the component tests. I'm not an expert in React Testing Library so if there are any tricks folks would like to share, I'm ready to learn!

Thank you for adding tests! I've also left a couple of suggestions there, but overall they're looking great!

File: packages/components/src/tools-panel/tools-panel-header/component.js
Comment (ramonjd) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: Thanks for thinking about this. I'll flip the verb back to "View" for now since I agree that it communicates more than "Show".

I'm fairly neutral on "sections". If folks think it describes the controls more succinctly/clearly I'm very willing to defer to those opinions. 👍

**Review by ciampo — APPROVED** [review]
Looking good! 🚀 

Thank you again for working on this and addressing the feedback!

**Comment by ramonjd** [comment]
Thanks for the continued feedback @ciampo ! 

I think we'll look at merging this one after after we do the [TS migration on the ToolsPanel](https://github.com/WordPress/gutenberg/pull/34028#pullrequestreview-756284323) to avoid causing more work for folks (including you) working on that 👍

File: packages/components/src/tools-panel/test/index.js
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/test/index.js]: I think we have a copy/paste error here.
```suggestion
		it( 'should render appropriate icons for the dropdown menu', async () => {
```

File: packages/components/src/tools-panel/test/index.js
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/test/index.js]: The default state of `controlProps` [should have a value](https://github.com/WordPress/gutenberg/blob/8a8987558044019e8c92eec916f91eff0e88646d/packages/components/src/tools-panel/test/index.js#L21) that would mean that it's actually "activated", as this test puts it, in the menu. If the test order is randomized or focused, this test fails. 

Some options for quick fixes might be switching `controlProps` to `altControlProps` which doesn't have a value. Better still is probably a before filter to ensure a known state of props. Or, just force an undefined value in the props used within the render call.

This might also be an issue elsewhere in these tests as well. There are plans to remove hardcoded CSS classes and improve these tests in a follow-up PR. So I think we can keep this proposed change to just this test. 

What do you think?

**Review by aaronrobertshaw — CHANGES_REQUESTED** [review]
Thank you @ramonjd for all the effort in putting this together and keeping it up-to-date while the ToolsPanel was continuing to be iterated on. 👍 

This tests nicely for me as per the instructions. I couldn't get the panel to display an unexpected icon.

I do wonder how this change will ultimately feel when multiple panels sit adjacent with different default controls and therefore potentially different toggle icons. Will the inconsistency across panels cause any confusion? The ToolsPanel itself is already adding to the inconsistency here, by differing from the chevron icon and the old panel.

I appreciate the idea is to help give users a nudge to open the panel to discover additional control options. My concern is that the `ToolsPanel` already has a high level of assumed knowledge in its use and does increasing inconsistency in the UI add to the confusion?

This isn't a show stopper of course. Also, there are plans to take a step back on the `ToolsPanel` and potentially rethink some of its design. This would give us a chance to leverage the lessons learnt and requirements discovered so far and improve this component greatly while it's still experimental.

There are a couple of tiny things I think we need to tweak on this then it should be good to go.

Thanks again for your patience on this one!

File: packages/components/src/tools-panel/test/index.js
Comment (ramonjd) [reply packages/components/src/tools-panel/test/index.js]: Thanks for spotting that.

File: packages/components/src/tools-panel/test/index.js
Comment (ramonjd) [reply packages/components/src/tools-panel/test/index.js]: Thanks a lot for pulling me up this. I'll test again and define a default state somehow, as you've suggested.

> There are plans to remove hardcoded CSS classes and improve these tests in a follow-up PR.

Testability without querying the DOM was one of the reasons why this PR introduced the aria label on the icon, so hopefully that's one less thing to change later 👍

**Comment by ramonjd** [comment]
Requested changes are in, thanks @aaronrobertshaw ! Just needs your approving review to unblock the PR.

> I do wonder how this change will ultimately feel when multiple panels sit adjacent with different default controls and therefore potentially different toggle icons. Will the inconsistency across panels cause any confusion? The ToolsPanel itself is already adding to the inconsistency here, by differing from the chevron icon and the old panel.

This thought also occurred to me. Ultimately it's something I think we can answer better when we use the ToolsPanel across more controls. 

I'd say that, in this case, it's okay to be conservative and delay introducing this PR's changes until we have more examples of the ToolsPanel in use in the wild. There is not a huge urgency on the plus icon. :)

Having said that, if the ToolsPanel is due for a UX review as you say, then it could be beneficial to have the plus icon already merged, and therefore available to consider in context when we do decide what to take beyond the experimental stage.

I understand the concern however that, once it's in, it's hard to remove. 

> My concern is that the ToolsPanel already has a high level of assumed knowledge in its use and does increasing inconsistency in the UI add to the confusion?

This might be our core issue. Until we solve the usability of the component, perhaps it is wise not to introduce more spice to the recipe.

**Comment by shaunandrews** [comment]
> there are plans to take a step back on the ToolsPanel and potentially rethink some of its design

Are there specific plans or a link you can share around this, @aaronrobertshaw ?

**Comment by aaronrobertshaw** [comment]
> Are there specific plans or a link you can share around this, @aaronrobertshaw ?

My apologies @shaunandrews, I have chosen my words poorly with that comment. Perhaps, "plans" was too strong a word and "design" covered more implementation 🙂 

There was some discussion ([one example](https://github.com/WordPress/gutenberg/pull/34530#pullrequestreview-753098588)) around the confusing dual-purpose nature of the panel's menu items and a feeling that there was too much assumed knowledge with the component. The only concrete plan I'm aware of at the moment is to vet this behaviour by actually using the panel. 

I've also been asked to iterate on the implementation of the `ToolsPanel`, including switching the panel to leverage the `Grid` component. This should be completed soon.

**Comment by ramonjd** [comment]
@aaronrobertshaw I think this one is ready for a final review. 

Personally I don't think it's so invasive a change, and it's one we can easily roll back. 

Nevertheless, I understand that there we are still working out the ToolsPanel UX and this could introduce an undesirable variable.

cc @mtias @shaunandrews for any final thoughts and preferences.

File: packages/components/src/tools-panel/stories/index.js
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/stories/index.js]: The current length of the label forces it to wrap so the menu placement looks odd. This little tweak means we can leave the panel width consistent with how it's used in the editor etc. What do you think?

```suggestion
					label="Tools Panel (optional items only)"
```

| Before | After |
|---|---|
| Screen Shot 2021-10-13 at 4 15 23 pm | Screen Shot 2021-10-13 at 4 15 06 pm |

File: packages/components/src/tools-panel/tools-panel/hook.ts
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/tools-panel/hook.ts]:
```
// Default state is that all optional controls, if any, are hidden.
```

I'm not sure the change of comment here is as clear as it could be. Optional controls are displayed when they contain a value.

```
// If no optional controls are present, then none are hidden and this will
// be `false`.
```

Also, is the change for this state to default to `true` needed? Semantically, I'd argue that if there are no controls then there are none hidden. Leaving this as `false` didn't appear to negatively impact the behaviour of this PR and allows us to not merge an unnecessary empty style for the panel.

File: packages/components/src/tools-panel/tools-panel/hook.ts
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/tools-panel/hook.ts]: Unless I'm missing something here the code here could stay as it is but we combine the comments to cover both cases. Perhaps something like:

```suggestion
	// Whether all optional menu items are hidden or not must be tracked
	// in order to later determine if the panel display is empty and handle
	// conditional display of a plus icon to indicate the presence of further 
	// menu items.
```

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for keeping this one up-to-date @ramonjd 👍 

The new `ToolsPanel` menu icon works as advertised for me.

✅  Tested via storybook, block editor and global styles
✅  `ToolsPanel` unit tests and type checking are all passing

I have a few minor comments or tweaks I think might be good to address. 

Then I think this is good to merge as you suggested. I don't believe it will interfere with anything around updating the `ToolsPanel` UX.

Thanks also for your patience on this one, I appreciate it's been delayed a few times because of other ongoing work to the `ToolsPanel. 🙇

File: packages/components/src/tools-panel/tools-panel/hook.ts
Comment (ramonjd) [reply packages/components/src/tools-panel/tools-panel/hook.ts]: Yeah, it's not clear. I like your comment suggestion below, much better thanks.

The default value and copy change was mostly due to a bad rebase.

I let dac782b6837bde1b03c5a58b7090913683822d3a through, instead of 2984e40cfc0cde894b2884a06cf99548942368fd

File: packages/components/src/tools-panel/tools-panel/hook.ts
Comment (ramonjd) [reply packages/components/src/tools-panel/tools-panel/hook.ts]: 🤦 

That this hook was [committed already 5 days ago](https://github.com/WordPress/gutenberg/commit/27cf9444ac5db0fce6184221a39ddd54dc37d0dc#diff-4f87334152aa083c77befd564ae2e8db3c95353c00d7e731abd2dbe883c5fdb5R112) threw me :D My brain short-circuited and didn't expect it to be in there in the rebase.

**Comment by ramonjd** [comment]
Thanks again for giving it the expert eye @aaronrobertshaw 

I've implemented the comment changes  – much clearer thank you – and have addressed the rebase anomalies 😇 

> Then I think this is good to merge as you suggested. I don't believe it will interfere with anything around updating the ToolsPanel UX.

I think I just need an official tick or a dismissal of your [last change request](https://github.com/WordPress/gutenberg/pull/34107#pullrequestreview-762721050) (I can't do it at my end) to be able to merge. Cheers!

**Review by aaronrobertshaw — APPROVED** [review]

**Comment by aaronrobertshaw** [comment]
I've given this a fresh test. Everything looks good to me 👍 

I'll leave the honours to you :shipit:

**Pull request merged**
